### PR TITLE
Adds multiplex sandbox support and improves argument parsing

### PR DESCRIPTION
**Pull request by jjudd**
Primary reviewer here is @jadenPete. I added @gregghz in case they're interested in taking a look.

### CODE REVIEW

File: rules/scala_proto/private/ScalaProtoWorker.scala [hunk 1]
```diff
@@ -1,31 +1,52 @@
...
 import net.sourceforge.argparse4j.impl.Arguments
-import net.sourceforge.argparse4j.inf.ArgumentParser
+import net.sourceforge.argparse4j.inf.{ArgumentParser, Namespace}
 import protocbridge.{ProtocBridge, ProtocRunner}
 import scala.jdk.CollectionConverters._
 import scalapb.ScalaPbCodeGenerator
 
 object ScalaProtoWorker extends WorkerMain[Unit] {
 
+  private class ScalaProtoRequest private (
```
Comment (jadenPete) [inline rules/scala_proto/private/ScalaProtoWorker.scala:21]: Nit: Should this be a `case class` (this applies to the other configuration classes created in this commit)?

File: src/main/scala/higherkindness/rules_scala/common/sandbox/SandboxUtil.scala [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+   *
+   * This is a convenience function for workers to easily get functions to prefix their path and file operations
+   */
+  def getSandboxPath(workDir: Path, path: Path): Path = {
+    workDir.resolve(path)
+  }
+  def getSandboxFile(workDir: Path, file: File): File = {
+    workDir.resolve(file.toPath).toFile
+  }
+  def getSandboxPaths(workDir: Path, paths: JList[Path]): List[Path] = {
```
Comment (jadenPete) [inline src/main/scala/higherkindness/rules_scala/common/sandbox/SandboxUtil.scala:25]: Nit: Do we need these `getSandboxPaths` methods? Writing a method for every collection seems a bit unnecessary when the caller could just as easily call `getSandboxPath` for every element.

File: rules/scalafmt/scalafmt/ScalafmtRunner.scala [hunk 3]
```diff
@@ -14,20 +18,38 @@ import scala.io.Codec
 
 object ScalafmtRunner extends WorkerMain[Unit] {
 
-  protected[this] def init(args: Option[Array[String]]): Unit = {}
+  private[this] class ScalafmtRequest private (
```
Comment (jadenPete) [inline rules/scalafmt/scalafmt/ScalafmtRunner.scala:21]: Nit: I think we should avoid using `private[this]` because it doesn't offer many guarantees over `private` and will be removed in Scala 3 (this applies to the other uses of `private[this]` and `protected[this]` in this commit.

File: src/main/scala/BUILD [hunk 4]
```diff
@@ -3,6 +3,7 @@ load(
     "configure_bootstrap_scala",
     "configure_zinc_scala",
     "scala_library",
+    # "bar_toolchain",
```
Comment (jadenPete) [inline src/main/scala/BUILD]: Can we remove this comment?

File: src/main/scala/BUILD [hunk 5]
```diff
@@ -67,17 +68,15 @@ configure_zinc_scala(
 
 # Scala 3
 
+# Adding this, so we make sure to have a Scala library in the
+# IntelliJ libraries, so we can get a Scala SDK on sync.
 # Adding this, so we make sure to have a Scala library in the
```
Comment (jadenPete) [inline src/main/scala/BUILD]: It looks like this comment was duplicated.

File: src/main/scala/BUILD [hunk 6]
```diff
@@ -67,17 +68,15 @@ configure_zinc_scala(
...
 # IntelliJ libraries, so we can get a Scala SDK on sync.
 scala_library(
     name = "scala-sdk",
```
Comment (jadenPete) [inline src/main/scala/BUILD:73]: Could you explain a bit more why this is needed? I have a couple questions:
- Does IntelliJ look in the classpath of generated JARs to find a Scala SDK upon sync?
  - If so, why do we need a bogus target like this when virtually every Scala target will have the standard library bundled with it?
- Why didn't we have a Scala 3 SDK before?

File: src/main/scala/higherkindness/rules_scala/common/args/ArgsUtil.scala [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+      case e: HelpScreenException =>
+        parser.handleError(e, new PrintWriter(out))
+        throw new AnnexWorkerError(0)
+      case e: ArgumentParserException =>
+        parser.handleError(e, new PrintWriter(out))
+        throw new AnnexWorkerError(1)
+    }
+  }
+
+  case class PathArgumentType() extends ArgumentType[Path] {
```
Comment (jadenPete) [inline src/main/scala/higherkindness/rules_scala/common/args/ArgsUtil.scala:30]: Nit: Shouldn't this be an `object` instead of a `case class`?

File: src/main/scala/higherkindness/rules_scala/common/args/ArgsUtil.scala [hunk 8]
```diff
@@ -0,0 +1,35 @@
+package higherkindness.rules_scala
```
Comment (jadenPete) [inline src/main/scala/higherkindness/rules_scala/common/args/ArgsUtil.scala:1]: Nit: Can we merge these two `package` statements into one? I always thought multiple `package` statements was a bad Scala feature.

File: src/main/scala/higherkindness/rules_scala/common/sandbox/SandboxUtil.scala [hunk 9]
```diff
@@ -0,0 +1,31 @@
+package higherkindness.rules_scala
```
Comment (jadenPete) [inline src/main/scala/higherkindness/rules_scala/common/sandbox/SandboxUtil.scala:1]: Nit: Can we merge these two `package` statements into one?

File: src/main/scala/higherkindness/rules_scala/common/sandbox/BUILD [hunk 10]
```diff
@@ -0,0 +1,14 @@
+load("//rules:scala.bzl", "scala_library")
+load("//rules:scalafmt.bzl", "scala_format_test")
+
+scala_library(
+    name = "sandbox",
+    srcs = glob(["**/*.scala"]),
```
Comment (jadenPete) [inline src/main/scala/higherkindness/rules_scala/common/sandbox/BUILD]: Nit: `glob(["*.scala"])`. We should do our best to prevent non-1:1:1d targets from cropping up.

File: src/main/scala/higherkindness/rules_scala/common/sbt-testing/Verbosity.scala [hunk 11]
```diff
@@ -0,0 +1,16 @@
+package higherkindness.rules_scala
```
Comment (jadenPete) [inline src/main/scala/higherkindness/rules_scala/common/sbt-testing/Verbosity.scala:1]: Nit: Can we merge these two `package` statements into one?

File: src/main/scala/higherkindness/rules_scala/common/sbt-testing/Verbosity.scala [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+package common.sbt_testing
+
+sealed abstract class Verbosity(val level: String)
+object Verbosity {
+  case object High extends Verbosity("HIGH")
+  case object Medium extends Verbosity("MEDIUM")
+  case object Low extends Verbosity("LOW")
+  val values = Map(
```
Comment (jadenPete) [inline src/main/scala/higherkindness/rules_scala/common/sbt-testing/Verbosity.scala:9]: Super nit:
```scala
val values = List(High, Medium, Low).map(value => value.level -> value).toMap
```

File: src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala [hunk 13]
```diff
@@ -13,11 +14,14 @@ trait WorkerMain[S] {
 
   protected[this] def init(args: Option[Array[String]]): S
 
-  protected[this] def work(ctx: S, args: Array[String], out: PrintStream): Unit
+  protected[this] def work(ctx: S, args: Array[String], out: PrintStream, workDir: Path): Unit
```
Comment (jadenPete) [inline src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala:17]: Wouldn't it be better for `workDir` to be an `Option[Path]`? Passing `""` seems pretty confusing.

File: src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala [hunk 14]
```diff
@@ -13,11 +14,14 @@ trait WorkerMain[S] {
...
+
+  protected[this] var isWorker = false
```
Comment (jadenPete) [inline src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala:19]: I'm not really a fan of using state like this because:
1. We have to remember to initialize it
2. Before we initialize it, it's inaccurate (i.e. it could be `false` when it should be `true`)

Can we pass this to `work` instead?

File: src/main/scala/higherkindness/rules_scala/workers/common/AnnexMapper.scala
Comment (jadenPete) [inline src/main/scala/higherkindness/rules_scala/workers/common/AnnexMapper.scala:1]: Could you fill me in on why these changes were necessary?

File: src/main/scala/higherkindness/rules_scala/workers/common/AnnexScalaInstance.scala [hunk 15]
```diff
@@ -52,11 +66,46 @@ object AnnexScalaInstance {
...
+    val key = workRequestJarToWorkerJar.values.mkString(":")
 
     Option(instanceCache.get(key)).getOrElse {
-      val instance = new AnnexScalaInstance(allJars)
+      // Copy all the jars to the worker's directory because in a sandboxed world the
+      // jars can go away after the work request, so we can't rely on them sticking around.
+      // This should only happen once per compiler version, so it shouldn't happen often.
+      workRequestJarToWorkerJar.foreach { case (workRequestJar, workerJar) =>
+        this.synchronized {
+          // Check for existence of the file just in case another request is also writing these jars
```
Comment (jadenPete) [inline src/main/scala/higherkindness/rules_scala/workers/common/AnnexScalaInstance.scala:98]: Is this check necessary with the `this.synchronized` block? If so, why?

File: src/main/scala/higherkindness/rules_scala/workers/common/CommonArguments.scala [hunk 16]
```diff
@@ -67,59 +102,93 @@ object CommonArguments {
...
+
+    parser
+  }
+
+  def apply(namespace: Namespace, workDir: Path): CommonArguments = {
+    val analysisArgs = Option(namespace.getList[JList[String]]("analysis")).map(_.asScala).getOrElse(List.empty)
+
+    val analyses: List[Analysis] = analysisArgs.map { analysisArg =>
+      // Analysis strings are of the format: _label analysis_store [jar ...]
+      val Buffer(label, analysisStore, jars @ _*) = analysisArg.asScala: @unchecked
```
Comment (jadenPete) [inline src/main/scala/higherkindness/rules_scala/workers/common/CommonArguments.scala]: Nit: Why the `@unchecked` here? Can we present a more informative error?

File: src/main/scala/higherkindness/rules_scala/workers/deps/DepsRunner.scala [hunk 17]
```diff
@@ -41,72 +104,81 @@ object DepsRunner extends WorkerMain[Unit] {
...
-      .fold[Seq[List[String]]](Nil)(_.asScala.toSeq.map(_.asScala.toList))
-      .map {
-        case label +: jars => label.tail -> jars.toSet
-        case _             => throw new Exception(s"Unexpected case in DepsRunner")
-      }
-    val labelToPaths = groups.toMap
-    def pathsForLabel(depLabel: String): Seq[String] = {
+    val label = workRequest.label
+    val directDepLabels = workRequest.directDepLabels
+    val groupLabelToJarPaths = workRequest.groups.map { group =>
```
Comment (jadenPete) [inline src/main/scala/higherkindness/rules_scala/workers/deps/DepsRunner.scala:117]: Nit: Can we inline `label` and `directDepLabels`? Also, should `groupLabelToJarPaths` be a `lazy val` on `DepsRunnerRequest`?

File: src/main/scala/higherkindness/rules_scala/workers/deps/DepsRunner.scala [hunk 18]
```diff
@@ -41,72 +104,81 @@ object DepsRunner extends WorkerMain[Unit] {
...
-      val usedWhitelist = namespace.getList[String]("used_whitelist").asScala.map(_.tail).toList
-      directLabels
-        .diff(usedWhitelist)
+    val readWriteMappers = AnnexMapper.mappers(workDir, isIncremental = false)
+    val readMapper = readWriteMappers.getReadMapper()
+    val usedPaths = Files
+      .readAllLines(workRequest.usedDepsFile)
+      .asScala
+      .view
+      // Use the read mapper on the used classpath entries in order to keep
```
Comment (jadenPete) [inline src/main/scala/higherkindness/rules_scala/workers/deps/DepsRunner.scala:142]: Out of curiosity, could you explain what the read mapper does?

File: src/main/scala/higherkindness/rules_scala/workers/zinc/compile/Deps.scala
Comment (jadenPete) [inline src/main/scala/higherkindness/rules_scala/workers/zinc/compile/Deps.scala:1]: Could you fill me in on why this changes was necessary?

File: src/main/scala/higherkindness/rules_scala/workers/zinc/compile/ZincRunner.scala [hunk 19]
```diff
@@ -62,99 +87,87 @@ object ZincRunner extends WorkerMain[Namespace] {
...
-  protected[this] def init(args: Option[Array[String]]) = {
+  protected[this] def init(args: Option[Array[String]]): ZincRunnerWorkerConfig = {
     val parser = ArgumentParsers.newFor("zinc-worker").addHelp(true).build
     parser.addArgument("--persistence_dir", /* deprecated */ "--persistenceDir").metavar("path")
     parser.addArgument("--use_persistence").`type`(Arg.booleanType)
     parser.addArgument("--extracted_file_cache").metavar("path")
     // deprecated
     parser.addArgument("--max_errors")
-    parser.parseArgsOrFail(args.getOrElse(Array.empty))
+    val namespace = parser.parseArgsOrFail(args.getOrElse(Array.empty))
```
Comment (jadenPete) [inline src/main/scala/higherkindness/rules_scala/workers/zinc/compile/ZincRunner.scala:97]: Should we use `ArgsUtil.parseArgsOrFailSafe` here?

File: src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala [hunk 20]
```diff
@@ -13,11 +14,14 @@ trait WorkerMain[S] {
 
   protected[this] def init(args: Option[Array[String]]): S
```
Comment (jadenPete) [inline src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala:15]: This method seems rarely used in subclasses of `WorkerMain` because argument parsing (which calls `ArgsUtil.parseArgsOrFailSafe`) now requires access to the output stream, which this method doesn't provide. Should we refactor it to provide that and make more places use `init` in a meaningful way?

**Review by jadenPete — APPROVED** [review]

File: rules/scala_proto/private/ScalaProtoWorker.scala (hunk 1 above)
Comment (jjudd) [reply rules/scala_proto/private/ScalaProtoWorker.scala:21]: I don't believe so because you get this error message related to Scala 3 migration:

```
src/main/scala/higherkindness/rules_scala/workers/common/CommonArguments.scala:15: error: access modifiers for `copy` method are copied from the case class constructor under Scala 3 (or with -Xsource-features:case-apply-copy-access)
Scala 3 migration messages are issued as errors under -Xsource:3. Use -Wconf or @nowarn to demote them to warnings or suppress.
Applicable -Wconf / @nowarn filters for this fatal warning: msg=<part of the message>, cat=scala3-migration, site=higherkindness.rules_scala.workers.common.CommonArguments
case class CommonArguments private (
           ^
1 error
```

File: src/main/scala/higherkindness/rules_scala/common/sandbox/SandboxUtil.scala (hunk 2 above)
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/common/sandbox/SandboxUtil.scala:25]: I originally was doing that, but found it to be enough boiler plate that I wrote these as convenience methods.

File: rules/scalafmt/scalafmt/ScalafmtRunner.scala (hunk 3 above)
Comment (jjudd) [reply rules/scalafmt/scalafmt/ScalafmtRunner.scala:21]: I think there's a performance advantage, however small, in Scala 2 to using `private[this]` vs `private` https://users.scala-lang.org/t/what-does-passing-this-after-access-modifier-mean/3838/5

> Something that might have a performance impact (though I think the JIT should be pretty good at eliminating this anyway), is that a private[this] val gets compiled to a field only. While private val gets compiled to a field and an accessor method.

File: src/main/scala/BUILD (hunk 4 above)
Comment (jjudd) [reply src/main/scala/BUILD]: Yep. Nice catch.

File: src/main/scala/BUILD (hunk 5 above)
Comment (jjudd) [reply src/main/scala/BUILD]: Yep. Fixing.

File: src/main/scala/higherkindness/rules_scala/workers/deps/DepsRunner.scala (hunk 17 above)
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/workers/deps/DepsRunner.scala:117]: I've inlined those two. I did not not change that val to be lazy due to the performance penalty of lazy and the locking that comes with it.

File: src/main/scala/higherkindness/rules_scala/workers/zinc/compile/Deps.scala
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/workers/zinc/compile/Deps.scala:1]: Normalize does this:

> Returns a path that is this path with redundant name elements eliminated.

I realized we weren't normalizing paths, so it would in theory be possible to get a path that points to the same spot, but the string doesn't match up. I wanted to eliminate that potential bit of non-determinism by normalizing the absolute paths.

File: src/main/scala/higherkindness/rules_scala/workers/zinc/compile/ZincRunner.scala (hunk 19 above)
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/workers/zinc/compile/ZincRunner.scala:97]: Not here, no. This is called with the worker is being initialized. It's ok for the worker to die here. If the worker dies here it just fails to start. It's also preferable for it to die as I'm not sure what else to do here.

If that worker dies on a work request, that's bad. Recovery is just telling Bazel that work request failed and then waiting for the next request to come in.

File: src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala (hunk 20 above)
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala:15]: I think this is still useful for when you need to initialize the worker.

`init` happens once when the worker starts. `work` happens on every work request.

Agree that it is called in frequently and a good tradeoff may be providing a default implementation of init, but I'm not sure off the top of my head how to accomplish that because of the generics.

File: src/main/scala/higherkindness/rules_scala/workers/deps/DepsRunner.scala (hunk 18 above)
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/workers/deps/DepsRunner.scala:142]: Read/write mapper is what we use to serialize and deserialize paths in a machine independent way.

If you didn't do mapping you'd end up with something like `/home/foo/project/src/scala/bar/Qux.scala`, but that's machine dependent. We only care about the `src/scala/bar/Qux.scala` bit. 

The write mappers write things to disk in a machine independent way.

The read mappers know how to read those things from disk and translate them to actual paths on your machine.

Most of this is defined in `src/main/scala/higherkindness/rules_scala/workers/common/AnnexMapper.scala`

File: rules/scala_proto/private/ScalaProtoWorker.scala (hunk 1 above)
Comment (jadenPete) [reply rules/scala_proto/private/ScalaProtoWorker.scala:21]: Oh, I think you need to remove the `private` constructor modifier too.

File: rules/scalafmt/scalafmt/ScalafmtRunner.scala (hunk 3 above)
Comment (jadenPete) [reply rules/scalafmt/scalafmt/ScalafmtRunner.scala:21]: Ah, good to know. This was just a nit, so I'm fine with keeping it either way.

File: src/main/scala/BUILD (hunk 6 above)
Comment (jjudd) [reply src/main/scala/BUILD:73]: Didn't have a Scala 3 SDK before because I forgot to add it.

The IntelliJ Scala + Bazel integration could use some improvements. Happy to chat more about this offline because I've got a fork of it that would work well with your toolchain work.

File: src/main/scala/higherkindness/rules_scala/workers/deps/DepsRunner.scala (hunk 18 above)
Comment (jadenPete) [reply src/main/scala/higherkindness/rules_scala/workers/deps/DepsRunner.scala:142]: That clarifies things. Thanks!

File: src/main/scala/higherkindness/rules_scala/workers/zinc/compile/Deps.scala
Comment (jadenPete) [reply src/main/scala/higherkindness/rules_scala/workers/zinc/compile/Deps.scala:1]: Gotcha. I wasn't sure if it was necessary as a consequence of introducing multiplexed, sandboxed workers, or if it was just good practice.

File: src/main/scala/higherkindness/rules_scala/workers/zinc/compile/ZincRunner.scala (hunk 19 above)
Comment (jadenPete) [reply src/main/scala/higherkindness/rules_scala/workers/zinc/compile/ZincRunner.scala:97]: Ah, that makes sense.

File: src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala (hunk 20 above)
Comment (jadenPete) [reply src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala:15]: I was mainly concerned with us re-parsing arguments in `work` in a lot of places. I wonder if removing the `args` parameter in `work` would help to enforce that we're doing argument parsing in `init`.

File: src/main/scala/higherkindness/rules_scala/common/args/ArgsUtil.scala (hunk 7 above)
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/common/args/ArgsUtil.scala:30]: I believe this needs to be a class and that argparse4j needs a new instance of the argument type. At least that's how the file argument type works https://github.com/argparse4j/argparse4j/blob/6c0f8590f7408025daa5f3b234e914cf805808a6/main/src/main/java/net/sourceforge/argparse4j/impl/Arguments.java#L275

File: src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala (hunk 20 above)
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala:15]: Argument parsing happens in work, though. In `work`, we parse the arguments for each work request.

In `init` we parse args for the worker as a whole. I think the only place we have a valid use case for this is the Zinc incremental compilation stuff, which I'd like to remove. Beyond that, I don't see a good reason to keep `init` around.

In another PR I can remove the incremental compilation piece and we should be able to remove init.

File: src/main/scala/higherkindness/rules_scala/common/args/ArgsUtil.scala (hunk 8 above)
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/common/args/ArgsUtil.scala:1]: It's done like this everywhere else in the codebase. I don't support it, but this is consistent with everywhere else. If we want to change it, I'm open to it, but let's keep consistency everywhere.

File: src/main/scala/higherkindness/rules_scala/common/sandbox/SandboxUtil.scala (hunk 9 above)
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/common/sandbox/SandboxUtil.scala:1]: Same response here.

File: src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala (hunk 20 above)
Comment (jadenPete) [reply src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala:15]: Ah, I was conflating command-line arguments vs. the arguments supplied with each worker task. Carry on.

File: src/main/scala/higherkindness/rules_scala/common/sandbox/BUILD (hunk 10 above)
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/common/sandbox/BUILD]: Good catch. Fixing.

We have a lot of these globs. If you wanna clean any of them up, go for it.

File: src/main/scala/higherkindness/rules_scala/common/sbt-testing/Verbosity.scala (hunk 11 above)
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/common/sbt-testing/Verbosity.scala:1]: Same response here.

File: src/main/scala/higherkindness/rules_scala/common/sbt-testing/Verbosity.scala (hunk 12 above)
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/common/sbt-testing/Verbosity.scala:9]: I like it, but I like the performance of the Map better because it avoids the `map` and then the `toMap`

File: src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala (hunk 13 above)
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala:17]: I thought about that, but went with a `Path` instead because:
 * Every work request has a work directory. It's either this directory `""` or a specific sandbox directory
 * You just move the weirdness to the other side with `.getOrElse("")` in many places..

File: src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala (hunk 14 above)
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala:19]: Challenge with passing it to work is that being a worker can be relevant outside the scope of a work request. It's an attribute that applies to the running process, not just a work request.

That's why I like handling it in the base class. It means others don't have to worry about it.

We already have some work that is done on init before we enter into our work loop, so it seemed to fit pretty well here.

File: src/main/scala/higherkindness/rules_scala/workers/common/AnnexMapper.scala
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/workers/common/AnnexMapper.scala:1]: Same reason as the other places I changed things to use Normalize

File: src/main/scala/higherkindness/rules_scala/workers/common/AnnexScalaInstance.scala (hunk 15 above)
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/workers/common/AnnexScalaInstance.scala:98]: I think the race condition happens when you have two separate AnnexScalaInstances being created that share a jar.

I'm likely being overly cautious because the Javadoc states "Copying a file is not an atomic operation."

https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html

I was worried about calling `Files.copy`, it not being atomic, and having the same file being copied from two different places at once and ending up with some funky corruption or something like that.

File: src/main/scala/higherkindness/rules_scala/workers/common/CommonArguments.scala (hunk 16 above)
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/workers/common/CommonArguments.scala]: Non-exhaustive match warning. I was doing unsafe things and wanted to do them unsafely.

```
src/main/scala/higherkindness/rules_scala/workers/common/CommonArguments.scala:167: error: match may not be exhaustive.
Applicable -Wconf / @nowarn filters for this fatal warning: msg=<part of the message>, cat=other-match-analysis, site=higherkindness.rules_scala.workers.common.CommonArguments.apply.analyses.$anonfun.x$2
      val Buffer(label, analysisStore, jars @ _*) = analysisArg.asScala
```

I've changed this to no longer use multiple assignment, which avoids the warning.

File: rules/scala_proto/private/ScalaProtoWorker.scala (hunk 1 above)
Comment (jjudd) [reply rules/scala_proto/private/ScalaProtoWorker.scala:21]: Agreed. I want the constructor to be private, though, which is why this is a class and not a case class.

I only want people to be able to construct these objects by passing in a namespace and a working directory, so someone doesn't accidentally construct one incorrectly.

File: src/main/scala/higherkindness/rules_scala/workers/zinc/compile/Deps.scala
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/workers/zinc/compile/Deps.scala:1]: Good point. This is unrelated to multiplex sandboxing. Just something I encountered while working on the sandboxing and applied everywhere. It really should have been its own commit. My bad.

File: rules/scala_proto/private/ScalaProtoWorker.scala (hunk 1 above)
Comment (jadenPete) [reply rules/scala_proto/private/ScalaProtoWorker.scala:21]: That makes sense. I think this is fine as-is, then.

File: src/main/scala/higherkindness/rules_scala/common/args/ArgsUtil.scala (hunk 8 above)
Comment (jadenPete) [reply src/main/scala/higherkindness/rules_scala/common/args/ArgsUtil.scala:1]: I didn't realize it was done like this elsewhere. Let's keep it for now, then.

File: src/main/scala/higherkindness/rules_scala/common/sandbox/BUILD (hunk 10 above)
Comment (jadenPete) [reply src/main/scala/higherkindness/rules_scala/common/sandbox/BUILD]: I'll probably do that in a later PR.

File: src/main/scala/higherkindness/rules_scala/common/sbt-testing/Verbosity.scala (hunk 12 above)
Comment (jadenPete) [reply src/main/scala/higherkindness/rules_scala/common/sbt-testing/Verbosity.scala:9]: Fair enough.

File: src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala (hunk 13 above)
Comment (jadenPete) [reply src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala:17]: Is `""` even a valid path though?

File: src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala (hunk 14 above)
Comment (jadenPete) [reply src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala:19]: That makes sense. I suppose we could refactor `WorkerMain` to accept the worker's command-line arguments and not override `main`, and then construct it inside the `main` method of the individual worker classes.

File: src/main/scala/higherkindness/rules_scala/workers/common/AnnexScalaInstance.scala (hunk 15 above)
Comment (jadenPete) [reply src/main/scala/higherkindness/rules_scala/workers/common/AnnexScalaInstance.scala:98]: I suppose we're better safe than sorry. I just thought the locking done by `this.synchronized` prevented this sort of thing.

File: src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala (hunk 13 above)
Comment (jjudd) [reply src/main/scala/higherkindness/rules_scala/common/worker/WorkerMain.scala:17]: Yeah, it ends up being the current working directory of the process, AFAIK. Which for a non-sandboxed worker is the correct working directory.

```
scala> Paths.get("").toAbsolutePath().toString()
val res3: String = /home/james
```
